### PR TITLE
Paris/interface tests

### PR DESCRIPTION
**Pull request by PhilippParis**

### CODE REVIEW

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/DAOException.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package com.at.ac.tuwien.sepm.ss15.edulium.dao;
+
+/**
+ * Exception indication an error in the persistence layer
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/DAOException.java]: indication -> indicating

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/MenuCategory.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * @param name sets the name of the category
+     */
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * @return  returns the identity if the object is persistent;
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/MenuCategory.java]: "... is persistent" isn't exactly true, the identity can be set (not null) even when the menu category isn't persistently stored (e.g. when using it for find by identity)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/MenuCategoryDAO.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * updates data of object menuCategory
+     * @param menuCategory object to update (must not be null)
+     * @throws DAOException if the object couldn't be updated
+     */
+    void update(MenuCategory menuCategory) throws DAOException;
+
+    /**
+     * removes the object menuCategory
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/MenuCategoryDAO.java]: ... from the underlying datasource

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/MenuCategoryDAO.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+     *
+     * sets the identity parameter of menuCategory, if the data was stored
+     * successfully
+     * @param menuCategory object to store (must not be null)
+     * @throws DAOException if the object couldn't be stored
+     */
+    void create(MenuCategory menuCategory) throws DAOException;
+
+    /**
+     * updates data of object menuCategory
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/MenuCategoryDAO.java]: ... in the underlying datasource

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/MenuCategory.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+     *          otherwise returns null
+     */
+    public Long getIdentity() {
+        return identity;
+    }
+
+    /**
+     * @param identity sets the identity of this object
+     *                 should only be called if the object is persistent
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/MenuCategory.java]: Can be called anytime (e.g. when using it for find by identity)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        menuCategoryDAO.create(cat);
+
+        // THEN
+        // check if identity is set
+        Assert.assertNotNull(cat.getIdentity());
+
+        // check retrieving object
+        MenuCategory matcher = new MenuCategory();
+        matcher.setIdentity(cat.getIdentity());
+        Assert.assertEquals(menuCategoryDAO.find(matcher), cat);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: find returns a list -> so store it and check if the size is 1, then check if first item matches

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    @Test(expected = DAOException.class)
+    public void testCreate_addingObjectWithoutNameShouldFail() throws DAOException {
+        // GIVEN
+        MenuCategory category = new MenuCategory();
+
+        // WHEN
+        try {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: remove try-catch (test expects the DAOException)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        }
+    }
+
+    @Test
+    public void testUpdate_shouldUpdateObject() throws DAOException {
+        // GIVEN
+        MenuCategory cat = new MenuCategory();
+        cat.setName("cat");
+
+        menuCategoryDAO.create(cat);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: Check if cat is stored (find by id + equals)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+        try {
+            menuCategoryDAO.create(category);
+        } finally {
+            Assert.assertNull(category.getIdentity());
+        }
+    }
+
+    @Test
+    public void testUpdate_shouldUpdateObject() throws DAOException {
+        // GIVEN
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: GIVEN -> PREPARE

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        MenuCategory cat = new MenuCategory();
+        cat.setName("cat");
+
+        menuCategoryDAO.create(cat);
+
+        // WHEN
+        cat.setName("newCat");
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: GIVEN

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        menuCategoryDAO.update(cat);
+
+        // THEN
+        // check if category name was updated
+        MenuCategory matcher = new MenuCategory();
+        matcher.setIdentity(cat.getIdentity());
+        Assert.assertEquals(menuCategoryDAO.find(matcher).get(0).getName(), "newCat");
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: find returns a list -> so store it and check if the size is 1, then check if first item matches

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    @Test(expected = DAOException.class)
+    public void testUpdate_updatingObjectWithIdentityNullShouldFail() throws DAOException {
+        // GIVEN
+        MenuCategory cat = new MenuCategory();
+        cat.setName("cat");
+
+        // WHEN
+        cat.setName("newCat");
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: can be removed

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+    @Test(expected = DAOException.class)
+    public void testUpdate_updatingNotPersistentObjectShouldFail() throws DAOException {
+        // GIVEN
+        MenuCategory cat = new MenuCategory();
+        cat.setName("cat");
+        cat.setIdentity(Long.valueOf(99999));
+
+        // WHEN
+        cat.setName("newCat");
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: can be removed

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+        menuCategoryDAO.update(cat);
+    }
+
+    @Test(expected = DAOException.class)
+    public void testUpdate_updatingNotPersistentObjectShouldFail() throws DAOException {
+        // GIVEN
+        MenuCategory cat = new MenuCategory();
+        cat.setName("cat");
+        cat.setIdentity(Long.valueOf(99999));
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: assure that no menu category with this id exists

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+        cat.setName("newCat");
+        menuCategoryDAO.update(cat);
+    }
+
+    @Test
+    public void testDelete_shouldDeleteObject() throws DAOException {
+        // GIVEN
+        MenuCategory cat = new MenuCategory();
+        cat.setName("cat");
+        menuCategoryDAO.create(cat);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: Check if cat is stored (find by id + equals)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+        // WHEN
+        menuCategoryDAO.delete(cat);
+
+        // check if category was removed
+        MenuCategory matcher = new MenuCategory();
+        matcher.setIdentity(cat.getIdentity());
+
+        Assert.assertEquals(menuCategoryDAO.find(matcher).size(), 0);
+        Assert.assertEquals(menuCategoryDAO.getAll().size(), 0);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: store the size of getAll() before adding the cat and check if the size is the same again after removing the cat

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+
+        // WHEN
+        menuCategoryDAO.delete(cat);
+    }
+
+    @Test(expected = DAOException.class)
+    public void testDelete_deletingNotPersistentObjectShouldFail() throws DAOException {
+        // GIVEN
+        MenuCategory cat = new MenuCategory();
+        cat.setIdentity(Long.valueOf(99999));
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: assure that no menu category with this id exists

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        cat3.setName("cat3");
+        menuCategoryDAO.create(cat1);
+        menuCategoryDAO.create(cat2);
+        menuCategoryDAO.create(cat3);
+
+        // WHEN
+        MenuCategory matcher = new MenuCategory();
+
+        matcher.setIdentity(cat1.getIdentity());
+        Assert.assertEquals(menuCategoryDAO.find(matcher).get(0), cat1);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: assure that the size of the list is 1

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+
+        matcher.setIdentity(cat2.getIdentity());
+        Assert.assertEquals(menuCategoryDAO.find(matcher).get(0), cat2);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: assure that the size of the list is 1

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+
+        matcher.setIdentity(cat3.getIdentity());
+        Assert.assertEquals(menuCategoryDAO.find(matcher).get(0), cat3);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: assure that the size of the list is 1

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+        List<MenuCategory> objects = menuCategoryDAO.find(matcher);
+
+        // THEN
+        Assert.assertEquals(objects.size(), 2);
+        Assert.assertTrue(objects.contains(cat1));
+        Assert.assertTrue(objects.contains(cat2));
+
+        // WHEN
+        matcher.setName(cat3.getName());
+        Assert.assertEquals(menuCategoryDAO.find(matcher).get(0), cat3);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: THEN and assure that the size is 1

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    @Test
+    public void testFind_shouldReturnEmptyList() throws DAOException {
+        //WHEN
+        MenuCategory matcher = new MenuCategory();
+        matcher.setIdentity(Long.valueOf(1));
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: assure that no menu category with this id exists

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java (hunk 7 above)
Comment (PhilippParis) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: there is no catch clause ;)
the test checks beforehand if the identity of the object is still NULL and then throws the exception

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java (hunk 10 above)
Comment (PhilippParis) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: ?

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java (hunk 7 above)
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestMenuCategoryDAO.java]: Ok nice :) Somehow missed the missing catch

**Pull request merged**